> **2026년 9월 코드 정비·재검증본**
> 4월 당시 원본과 출력은 `archive/notebooks/`에 보존했습니다. 아래 수치는 당시 성과를 소급 수정한 것이 아니라 새 평가 절차로 측정한 결과입니다.
> 저장된 결과를 확인하는 노트북입니다. 전체 재학습: `python -m reanalysis.pipeline` 및 `python -m reanalysis.legacy_shared`.

# 03. 팀 공통 저장 모델의 추가 학습 재검증
개인 최고 모델을 직접 제공했다고 주장하지 않습니다. 팀원 결과를 비교해 공통 모델을 정하고 추가 분석을 진행한 협업 단계입니다. 여기서는 실제 제공된 `base_model.keras`를 사용합니다.

원본 인코더·결측치 처리 객체는 보존되지 않아 survey와 원본 코드로 복원했습니다. 저장 scaler와 재구성 scaler의 수치 일치를 확인했지만, 모델의 전체 학습 이력까지 증명하지는 못합니다.

In [1]:
from pathlib import Path
import pandas as pd
import json
ROOT = Path.cwd()
RESULTS = ROOT / 'reanalysis' / 'results'
assert RESULTS.exists(), '저장소 루트에서 실행하세요.'

In [2]:
r = json.loads((RESULTS / 'legacy_shared_results.json').read_text(encoding='utf-8'))
print('scaler match:', r['scaler_matches_original_split'])
print('selected on validation:', r['selected_by_validation'])
print(pd.read_csv(RESULTS / 'legacy_shared_metrics.csv').to_string(index=False))

scaler match: True
selected on validation: Continue_all_layers
                    model  partition  accuracy  macro_f1  recall_0  recall_1   n  threshold
Original_shared_no_update validation  0.923529  0.923400  0.881720  0.974026 170        0.5
Original_shared_no_update       test  0.905714  0.905714  0.832461  0.993711 350        0.5
Scratch_same_architecture validation  0.929412  0.928421  0.956989  0.896104 170        0.5
Scratch_same_architecture       test  0.882857  0.881533  0.905759  0.855346 350        0.5
      Continue_all_layers validation  0.952941  0.952521  0.956989  0.948052 170        0.5
      Continue_all_layers       test  0.917143  0.916647  0.910995  0.924528 350        0.5
           Fine_tune_head validation  0.935294  0.935022  0.913978  0.961039 170        0.5
           Fine_tune_head       test  0.922857  0.922715  0.884817  0.968553 350        0.5


## 수정한 핵심
- 변환할 때마다 중앙값/인코딩을 다시 학습하지 않습니다.
- 78개 입력 이름과 순서를 저장된 scaler에 맞춥니다.
- 추가 학습 실험마다 원본 모델을 독립적으로 다시 불러옵니다.
- 모델별 예측을 다시 계산합니다.
- 추가 학습 850건을 680/170건으로 나누고 검증 기준을 먼저 정합니다.
- 350건 평가는 이미 과거에 살펴본 데이터이므로 새 블라인드 벤치마크라고 표현하지 않습니다.

In [3]:
import inspect
from reanalysis.legacy_shared import LegacyPreprocessor
print(inspect.getsource(LegacyPreprocessor))

class LegacyPreprocessor:
    def __init__(self, survey, scaler):
        self.scaler = scaler
        self.medians = survey[RATINGS + CONTINUOUS].median()
        self.modes = survey[CATEGORICAL].mode().iloc[0]
        self.mapping = {c: {v: i for i, v in enumerate(sorted(survey[c].dropna().unique()))} for c in CATEGORICAL}
        self.columns = list(scaler.feature_names_in_)

    def frame(self, df):
        data = df[FEATURES].copy()
        data[RATINGS + CONTINUOUS] = data[RATINGS + CONTINUOUS].fillna(self.medians)
        data[CATEGORICAL] = data[CATEGORICAL].fillna(self.modes)
        for col in CATEGORICAL:
            data[col] = data[col].map(self.mapping[col])
            if data[col].isna().any():
                raise ValueError(f'Unknown legacy category: {col}')
        for col in RATINGS:
            if not data[col].isin(range(6)).all():
                raise ValueError(f'Invalid legacy rating: {col}')
            data[col] = pd.Categorical(data[col], categories=range(

## 해석 한계
단일 seed·작은 표본의 재검증입니다. 추가 학습이 항상 개선을 보장하지 않으며, 모델 성능 변화가 서비스 개선 효과를 의미하지 않습니다. 원본 파일은 덮어쓰지 않았습니다.